components/script_layout_interface/wrapper_traits.rs

```diff
@@ -183,22 +183,129 @@ pub trait ThreadSafeLayoutNode: Clone + Copy + NodeInfo + PartialEq + Sized {
         }
     }
 
+    fn get_before_pseudo(&self) -> Option<Self> {
+        self.as_element().and_then(|el| el.get_before_pseudo()).map(|el| el.as_node())
+    }
+
+    fn get_after_pseudo(&self) -> Option<Self> {
+        self.as_element().and_then(|el| el.get_after_pseudo()).map(|el| el.as_node())
+    }
+
+    fn get_details_summary_pseudo(&self) -> Option<Self> {
+        self.as_element().and_then(|el| el.get_details_summary_pseudo()).map(|el| el.as_node())
+    }
+
+    fn get_details_content_pseudo(&self) -> Option<Self> {
+        self.as_element().and_then(|el| el.get_details_content_pseudo()).map(|el| el.as_node())
+    }
+
     fn debug_id(self) -> usize;
 
     /// Returns an iterator over this node's children.
     fn children(&self) -> LayoutIterator<Self::ChildrenIterator>;
 
-    /// If this is an element, accesses the element data. Fails if this is not an element node.
+    /// Returns a ThreadSafeLayoutElement this is an element, None otherwise.
     #[inline]
-    fn as_element(&self) -> Self::ConcreteThreadSafeLayoutElement;
+    fn as_element(&self) -> Option<Self::ConcreteThreadSafeLayoutElement>;
+
+    #[inline]
+    fn get_pseudo_element_type(&self) -> PseudoElementType<Option<display::T>> {
+        self.as_element().map_or(PseudoElementType::Normal, |el| el.get_pseudo_element_type())
+    }
+
+    fn get_style_and_layout_data(&self) -> Option<OpaqueStyleAndLayoutData>;
+
+    fn style(&self, context: &SharedStyleContext) -> Arc<ServoComputedValues> {
+        if let Some(el) = self.as_element() {
+            el.style(context)
+        } else {
+            debug_assert!(self.is_text_node());
+            self.style_for_text_node()
+        }
+    }
+
+    fn selected_style(&self) -> Arc<ServoComputedValues> {
```

Comment:
Yeah I was kinda on the fence about it, but decided to opt for removal, since we can always add it back if we need it.